Python 4 Problem Set -- Lists and Loops
===================

1. For the next series of tasks about lists use the interpreter:  
	a.  Create a list of 5 of your favorite things. 

```python

>>> my_fave_things = ['pizza', 'music', 'books', 'neature', 'swimming']              
```
 
	b.  Use the `print()` function to print your list.

```python
>>> print(my_fave_things)
['pizza', 'music', 'books', 'neature', 'swimming'] 
```
  
	c.  Use the `print()` function to print out the middle element.

```python
>>> print(my_fave_things[2])                                                        
books
```
  
	d.  Now replace the middle element with a different item, your favorite song, or song bird.

```python
>>> my_fave_things[2] = 'magpie'
```
  
	e.  Use the same print statement from b. to print your new list. Check out the differences.

```python                                                 
>>> print(my_fave_things)                                                            
['pizza', 'music', 'magpie', 'neature', 'swimming']
``` 
   
	f.  Add a new element to the end. [Read about append()](https://www.tutorialspoint.com/python/list_append.htm).  

```python
>>> my_fave_things.append('lotus') # append adds element to the end of lists
>>> my_fave_things
['pizza', 'music', 'magpie', 'neature', 'swimming', 'lotus']   
```

	g.  Add a new element to the beginning. [Read about insert()](https://www.tutorialspoint.com/python/list_insert.htm).  

```python
>>> my_fave_things.insert(0, 'swimming') # insert adds an element at the specified index
>>> my_fave_things
['swimming', 'pizza', 'music', 'magpie', 'neature', 'swimming', 'lotus'] 

```

        h.  Add a new element somewhere other than the beginning or the end.  

```python
>>> my_fave_things.insert(1, 'cookies') # insert adds an element at the specified index
>>> my_fave_things
['swimming','cookies', 'pizza', 'music', 'magpie', 'neature', 'swimming', 'lotus'] 

```

        i.  Remove an element from the end. [Read about pop()](https://www.tutorialspoint.com/python/list_pop.htm).  

```python
>>> my_fave_things = ['pizza','cookies','music', 'magpie', 'neature', 'swimming', 'lotus']
>>> my_fave_things.pop() # pop() removes the last element in a list and returns that element
'lotus'
>>> my_fave_things
['pizza', 'cookies', 'music', 'magpie', 'neature', 'swimming'] 
```       

	j.  Remove an element from the beginning.  

```python
>>> my_fave_things
['pizza', 'cookies', 'music', 'magpie', 'neature', 'swimming']
>>> my_fave_things.pop(0) # if an element is specified for pop(), then that value is returned and removed from the list
'pizza'
>>> my_fave_things
['cookies', 'music', 'magpie', 'neature', 'swimming']  
```

        k.  Remove an element from somewhere other than the beginning or the end.  

```python
>>> my_fave_things
['cookies', 'music', 'magpie', 'neature', 'swimming']
>>> my_fave_things.pop(3)
>>> my_fave_things
['cookies', 'music', 'magpie', 'swimming']
```

	l.  Use `join` to create a string. Join the elements on ', ' 
	
```python
>>> my_fave_things_string = ', '
>>> my_fave_things_string.join(my_fave_things)  ## join is a string method which will concentenate the elements of a specified list using the string as a separator
'cookies, music, magpie, swimming'
```


2. Write a script that splits a string into a list. In your script: 
	- Save the string `sapiens, erectus, neanderthalensis` as a variable named `taxa`.
	- Print `taxa`.
	- Print `taxa[1]`, what do you get?
	- Print `type(taxa)`. What is it's type?
	- Split `taxa` into individual words and print the result of the split. (Think about the ', '.)
	- Store the result of `split` in a new variable named `species`.
	- Print `species`.
	- Print the `species[1]`, What do you get?
	- Print `type(species)`. What is it's type?
	- Sort the list alphabetically and print (hint: lookup the function `sorted()`). 
	- Sort the list by length of each string and print. (The shortest string should be first). [Check out documentation of the key argument](https://www.programiz.com/python-programming/methods/built-in/sorted).

```python
#!/usr/bin/env python3

taxa = 'sapiens, erectus, neanderthalensis' ## saves the specified string as taxa

print(taxa)  ## prints the variable taxa
print(taxa[1]) ## prints the first index (the second character) in taxa (a string)
print(type(taxa))  ## prints the type of the variable taxa

species = taxa.split(', ')  ## split is a string method which returns a list generated by separating a string, using the specified input

print(species)  ## prints the variable species (which looks similar, but different to how taxa appeared when printed)
print(species[1])  ## prints the first index (the second item) in species (a list)
print(type(species))  ## prints the type of the variable species

print(species)
sorted_species = sorted(species) ## will return a sorted version of the list, without changing the order of the list

print(species)
print(sorted_species)

species.sort() ## will sort the list, actually changing the list

print(species)

re_sorted_species = sorted(species, key = len) ## sorted can be fed a key to sort by, in this case length (remember sorted doesn't change the underlying list)

print(re_sorted_species)
print(species)

```

The script returns the following (I have added notes)

```python
sapiens, erectus, neanderthalensis		## the taxa string printed
a						## the second character from taxa
<class 'str'>					## the type of taxa is printed
['sapiens', 'erectus', 'neanderthalensis']	## species, the list, printed
erectus						## the second element of species
<class 'list'>					## the type of species is printed
['sapiens', 'erectus', 'neanderthalensis']	## species is printed to show order
['sapiens', 'erectus', 'neanderthalensis']	## species is printed, after sorted() is used to show you that it has not changed
['erectus', 'neanderthalensis', 'sapiens']	## the list generated using sorted() is printed
['erectus', 'neanderthalensis', 'sapiens']	## species is printed after applying the sort method
['erectus', 'sapiens', 'neanderthalensis']	## species are sorted by length using sorted()
['erectus', 'neanderthalensis', 'sapiens']	## species is reprinted to highlight that sorted() did not change the alphabetically sorted species
```


3. Using the Python interpreter, interrogate the difference between these two ways to copy a list. Careful! One of these is NOT what you might expect. 
   - Method 1
     - Create a list. For example: `my_list = ['a', 'bb', 'ccc']`
     - Make a copy using the `=` assignment operator:  `list_copy = my_list`
     - Print the original list `print(my_list)`
     - Alter the `list_copy` by adding a new element using `append()`
     - Print the original list again `print(my_list)`
   - Method 2  
     - Create a list. For example: `my_list2 = ['a', 'bb', 'ccc']`  
     - Make a copy with the copy() method `list_copy2 = my_list2.copy()`  
     - Print the original list `print(my_list2)`
     - Alter the `list_copy2` by adding a new element using `append()`   
     - Print the original list again `print(my_list2)`   

```python
>>> my_list = ['a', 'bb', 'ccc']
>>> your_list = my_list
>>> print(my_list)
['a', 'bb', 'ccc']
>>> your_list.append('dddd')
>>> print(my_list)
['a', 'bb', 'ccc', 'dddd']
>>> print('Woooooooaaaaaaaahhhhhh')	## changing your_list changed my_list - I'm hecking shook
Woooooooaaaaaaaahhhhhh

>>> my_list2 = ['1', '22', '333']
>>> my_list2_copy = my_list2.copy()
>>> print(my_list2)
['1', '22', '333']
>>> my_list2_copy.append('4444')
>>> print(my_list2)			## a copy made using the .copy() method keeps things nice and separated
['1', '22', '333']
>>> print(my_list2_copy)
['1', '22', '333', '4444']

```

4. Write a script that uses a `while` loop to print out the numbers 1 to 100.

```python
#!/usr/bin/env python3

count = 1                       ## initialize the counter

while count < 101:              ## specifies the conditions for the while loop
	print(count)            ## prints the active count variable
	count += 1              ## increments the counter by 1 
```

5. Write a script that uses a `while` loop to calculate the [factorial](https://en.wikipedia.org/wiki/Factorial) of 1000. 

```python
#!/usr/bin/env python3

count = 1000           			## initialize the counter
fact_product = 1                        ## initilizes the factorial product

while count > 0:                	## specifies the conditions for the while loop
	fact_product *= count   	## performs the multiplicaiton required to calculate the factorial
	count -= 1              	## increments the counter by 1
	print(fact_product)             ## a very, very large number (note this is not in theloop, so that it only prints once)
```
The factorial starts with 402387260..........

6. Iterate through each element of this list using a `for` loop: [101,2,15,22,95,33,2,27,72,15,52]
   - Print out only the values that are even (hint: use the modulus operator).
```python
#!/usr/bin/env python3

the_q6_list = [101,2,15,22,95,33,2,27,72,15,52] ## the list from the prompt

for value in the_q6_list:                       ## sets up the for loop
	if value%2 ==0:                         ## tests whether the value is even
		print(value)                    ## prints only if even
```
   
7. Sort the elements of the above list, then iterate through each element using a `for` loop and:
   - Print each element.
   - Calculate two cumulative sums, one of all the even values and one of all the odd values. 
   - Print only the final two sums. The output from your script should be:
   
   ```
   Sum of even numbers: 150 
   Sum of odds: 286
   ```
```python
#!/usr/bin/env python3

the_q6_list = [101,2,15,22,95,33,2,27,72,15,52] ## the list from the prompt

the_q6_list.sort()                              ## use the .sort() method to sort the list

even_sum = 0                                    ## initializes even sum
odd_sum = 0                                     ## initializes odd sum

for value in the_q6_list:                       ## sets up the for loop
	if value%2 ==0:                         ## tests whether the value is even
		even_sum += value               ## increments even sum by value
	else:                                   ## not even values (i.e. odd) are parsed to here
		odd_sum += value                ## increments odd sum by value
	print(value)                            ## value is printed outside of the if statement, such that it is printed no matter what

print('Even sum is:',even_sum)			## values are printed outside the loopso that they are only printed once and are the final value
print('Odd sum is:',odd_sum)

```
8. Write a script that uses `range()` in a `for` loop to print out every number between 0 and 99  
      - Modify your loop to print out every number between 1 and 100.

```python
#!/usr/bin/env python3

for number in  range(100):      ## specifies the conditions for the while loop
	print(number)           ## prints number, the active variable
	print(number + 1)       ## increments the active variable by 1 to shift printed values from 0-99 to 1-100
```
      
9. Create a new script that uses list comprehension to do the same thing as problem 8. (Use `range()` to print out every number between 1 and 100.)

```python
#!/usr/bin/env python3

print([x+1 for x in range(100)])                ## prints the list generated using list comprehension -- recall range starts at zero and goes up to, but does not include the value that you specify  
```
 
10. Write a new script that takes the start and end values from the command line. If you call your script like this `count.py 3 10` it will print the numbers from 3 to 10.
      - Modify your script so that it will only print the number if it is odd.

```python

#!/usr/bin/env python3

import sys

start_val = int(sys.argv[1])            ## import the start and end values from standard in
end_val = int(sys.argv[2])

for val in range(start_val, end_val+1): ## specifies the range of the for loop
	print(val)                      ## prints the value specified by the range

for val in range(start_val, end_val+1): ## specifies the range of the for loop
	if val%2 !=0:                   ## evalatues to True if val is odd
		print(val)                                                            
```
      
11. Write a new script to create a list with the following data  `['ATGCCCGGCCCGGC','GCGTGCTAGCAATACGATAAACCGG', 'ATATATATCGAT','ATGGGCCC']`. Use a `for` loop to iterate through each element of this list and:
   - Print out each element.
   - Print out the length and the sequence, separated by a tab. The output should look like:
   
   ```
   14	ATGCCCGGCCCGGC
   25	GCGTGCTAGCAATACGATAAACCGG
   12	ATATATATCGAT
   8	ATGGGCCC
   ```
```python

#!/usr/bin/env python3

nt_list = ['ATGCCCGGCCCGGC','GCGTGCTAGCAATACGATAAACCGG', 'ATATATATCGAT','ATGGGCCC'] 

for nt in nt_list:                      ## iterates through each value of list
	print(nt)

for nt in nt_list:
	print(len(nt),nt,sep='\t')      ## prints length and nt in list, separated by tab
```

12. Use list comprehension to generate a list of tuples. The tuples should contain sequences and lengths from the previous problem. Print out the length and the sequence (i.e., "4\\tATGC\\n").  A list of tuples looks like this [(4,'ATGC'),(2,'GC')]. Use `yourList.append( yourTuple )`.  Check out the second example [here](https://www.w3schools.com/python/ref_list_append.asp)


```python

#!/usr/bin/env python3 

nt_list = ['ATGCCCGGCCCGGC','GCGTGCTAGCAATACGATAAACCGG', 'ATATATATCGAT','ATGGGCCC']

my_tup_list = [(len(nt),nt) for nt in nt_list]		## generates a list of tuples with form (length of sequence, sequence)


```


13. Modify the script from #11 so that you also print out the number (postion in the list) of each sequence (i.e., "1\\t4\\tACGT\\n")

```python

#!/usr/bin/env python3

nt_list = ['ATGCCCGGCCCGGC','GCGTGCTAGCAATACGATAAACCGG', 'ATATATATCGAT','ATGGGCCC']

for index in range(len(nt_list)):       ## specifies the iterator as an index as opposed to iterating through the list using 'Python-magic'
	print(index + 1, len(nt_list[index]), nt_list[index],sep='\t')  ## prints index (scaled by 1 for humans), length and nt in list, separated by tab

```

14. Have you been commiting you work?

## Fun challenge problems! These are real scripts you might use in real life. You have already learned all you need to know to do each. If you don't have enough time in this session to complete, come back and try later.

1. Create a shuffled sequence ([Fisher-Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle))
    - Use a `for` loop to perform the following procedure N times (N = length of seq)
    - Select a random position A with `randrange()`
    - Select a random position B with `randrange()`
    - Exchange the letters at list indices A and B
    - Print the final shuffled sequence
    - Remember to test your code with test data. 

```python

#!/usr/bin/env python3

import random						## imports the random module

dna = input('Please enter a DNA sequence: ')		## prompts the user for a DNA sequence, saving as a string
dnaLen = len(dna)					## records the len of the DNA sequence
dnaLis = list(dna)					## converts the string to a list to enable site-specific mutation

for nt in range(dnaLen):				## sets up for loop to go site by site through sequence
   randApos=random.randrange(dnaLen)			## select two random positions
   randBpos=random.randrange(dnaLen)
   randA = dnaLis[randApos]				## record the nt at the specified random positions
   randB = dnaLis[randBpos]
   dnaLis[randApos] = randB				## swap out the nt at those positions
   dnaLis[randBpos] = randA
   print(dnaLis)					## not necessary, but lets you view sequence as it is shuffled	

dnaSeq=''						## initialize empty string
for nt in range(dnaLen):				## a for loop to generate a string version of dna sequence from its list representation
   dnaSeq += dnaLis[nt]
   
print(dnaSeq)

```



2. Calculate sequence identity: Start with 2 very similar DNA sequences. Use your favorites or use [Python_04.fasta](https://raw.githubusercontent.com/prog4biol/pfb2019/master/files/Python_04.fasta)
    - Align with ClustalW, TCoffee, or some other web alignment application. 
    - Output should be in FASTA format.
    - Store (copy and paste) each aligned sequence, including dashes, as two separate string variables. 
    - Get rid of newlines (if any). Newline characters are not part of sequence!
    - Use a `for` loop with `range()` to compare each index for nucleotide differences.
    - Report percent identity of the two sequences.

3. A new Restriction Fragments script:
   - Find [EcoRI](https://www.neb.com/products/r0101-ecori#Product%20Information) in this DNA sequence
```
GATGGGATTGGGGTTTTCCCCTCCCATGTGCTCAAGACTGGCGCTAAAAGTTTTGAGCTTCTCAAAAGTCTAGAGCCACCGTCCAGGGAGCAGGTAGCTGCTGGGCTCCGGGGACACTTTGCGTTCGGGCTGGGAGCGTGCTTTCCACGACGGTGACACGCTTCCCTGGATTGGCAGCCAGACTGCCTTCCGGGTCACTGCCATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATTCGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGTGGGGTTTTCCCCTCCCATGTGCTCAAGACTGGCGCTAAAAGTTTTGAGCTTCTCAAAAGTCTAGAGCCACCGTCCAGGGAGCAGGTAGCTGCTGGGCTCCGGGGACACTTTGCGTTCGGGCTGGGAGCGTGCTTTCCACGACGGTGACACGCTTCCCTGGATTGGCAGCCAGACTGCCTTCCGGGTCACTGCCATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATTCGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGAC
```
   - replace the EcoRI site 'GAATTC' with this 'G^AATTC'
   - split the new formatted sequence on the cut sites, store the resulting fragments in a list
   - iterate over each fragment and report
      - the start position in the original sequence
      - the end postion in the orginal sequence
      - the length of each fragemnt
   - sort the fragments by length and print out as they would appear on a agrose gel. (big to little)
